atef/archive_device.py

```diff
@@ -0,0 +1,732 @@
+
+    new_class = type(
+        f"{class_prefix}{cls.__name__}",
+        (cls, *new_bases),
+        clsdict
+    )
+    cache[cls] = new_class
+    logger.debug("cache[%s] = %s", cls, new_class)
+    return new_class
+
+
+class ArchiverPV(PyepicsPvCompatibility):
+    """
+    An epics.PV-like interface to archiver appliance data.
+
+    Notes
+    -----
+    There is a 1-to-1 correspondence of :class:`ArchiverPV` to Component,
+    whereas normally :class:`epics.PV` can be shared.
+    """
+
+    # `_make_connection` below will be in the main thread
+    _connect_in_thread: ClassVar[bool] = False
+
+    archive_timestamp: datetime.datetime = datetime.datetime.now()
+
+    def _make_connection(self):
+        """PyepicsPvCompatibility hook at startup."""
+        helper = ArchiverHelper.instance()
+        helper.queue_pv(
+            pvname=self.pvname,
+            dt=self._referrer_timestamp,
+            callback=self._archiver_initial_data,
+        )
+
+    def _archiver_initial_data(self, data: ArchivedValue):
+        """ArchiverHelper.queue_pv callback."""
+        # found_in_archiver = data.appliance is not None
+        self._update_state_from_archiver(data)
+        self._change_connection_status(connected=True)
+
+    @property
+    def _referrer_timestamp(self) -> datetime.datetime:
+        """
+        Timestamp from the referrer in the following order of precedence:
+        1. Timestamp from the _root_ device in the signal's device hierarchy.
+        2. The immediate parent device's timestamp.
+        3. This PV instance's timestamp (final fallback; should not be hit)
+        """
+        try:
+            return self._referrer.root.archive_timestamp
+        except AttributeError:
+            try:
+                return self._referrer.parent.archive_timestamp
+            except AttributeError:
+                return self.archive_timestamp
+
+    def put(
+        self,
+        value: Any,
+        wait: bool = False,
+        timeout: float = 30.0,
+        use_complete: bool = False,
+        callback: Optional[PyepicsPutCallback] = None,
+        callback_data: Optional[Any] = None,
+    ):
+        """Stub for pv.put. Callback will be run, but no change will be made."""
+        logger.warning(
+            "This is archived mode; no puts are allowed. "
+            "Attempted to change %r to %r",
+            self.pvname,
+            value,
+        )
+        return super().put(
+            value,
+            wait=wait,
+            timeout=timeout,
+            use_complete=use_complete,
+            callback=callback,
+            callback_data=callback_data,
+        )
+
+    def _update_state_from_archiver(
+        self, data: ArchivedValue, as_string=None
+    ) -> Dict[str, Any]:
+        """Update _args and run callbacks based on the ArchivedValue."""
+        if data.value is None:
+            as_string = as_string if as_string is not None else self.as_string
```

Comment:
Is as_string only used when there is no data? That seems wrong to me, unless I misunderstand the intent here.

Reply:
Tried to adjust this in 2ce341b; it should indeed be used in both scenarios. I also made some comments regarding how it would be more appropriate to mark it as "disconnected" if not in the archiver, but I have a feeling that will annoy us with broken devices that have a single non-archived PV.